### ISSUE TITLE
UI for visualizing the merge queue

### ISSUE THREAD

**Issue by scriptnull** [chdsbd/kodiak#616]
Currently, there is no way to see what is currently in the merge queue of kodiak without having to visit the pull requests one by one. It would be nice to have a section in the dashboard from where we could visualize this information.

**Comment by chdsbd** [comment]
Thanks the request.

Would you take action based on the merge queue positions or are you just curious to see the positioning?
Also, did you imagine the page would update in real time? I think this feature would be simpler to implement if the page didn't update live.

I think we have enough information in Redis to display tables like the following:

---

**chdsbd/kodiak/main**
position|pull request|added
---|---|---
1|[543](#example)|10 minutes ago
2|[545](#example)|5 minutes ago
3|[546](#example)|4 minutes ago

**chdsbd/kodiak/dev**
position|pull request|added
---|---|---
1|[234](#example)|3 minutes ago
2|[250](#example)|1 minute ago
3|[255](#example)|1 minute ago

---

Merge queues are per-repository, per-target and stored in Redis. So each branch (e.g. "main" and "dev") has a merge queue. The merge queue names are defined by the following code:

https://github.com/chdsbd/kodiak/blob/871cf119cd888dd7bbdf2e763761a53030e5a868/bot/kodiak/queue.py#L319-L321

To list all the merge queue names we could probably do the following, given an installation ID of 1123902:

```
SCAN 0 MATCH merge_queue:1123902* COUNT 20
```

The merge queues are sorted sets with the rank being the timestamp the event was added to the queue. The add command looks like:

```
ZADD merge_queue:1123902.sbdchd/squawk/main 1614997353.8109288 {"repo_owner": "sbdchd", "repo_name": "squawk", "pull_request_number": 534, "installation_id": "1923012", "target_name": "main"}
```

We can use `ZRANK` to list all the pull requests in a queue:

```
ZRANK merge_queue:1123902.sbdchd/squawk/main 0 1000
```


With the `SCAN` and `ZRANK` we should be able to make a page of merge queue tables.

**Comment by scriptnull** [comment]
> Would you take action based on the merge queue positions or are you just curious to see the positioning?

Our primary use case is when a person's PR is queued way behind in the queue (lets say 4th position), they tend to ask what are the PRs that are before their PR in-order to tentatively figure out how much time will those PRs will take. Having this feature answers this question without having to manually scan through all PRs with the auto-merge label.

It would also be perfect for helping us in introspecting how can we make our CI builds better in-order for the PRs in the queue to be merged faster at a given point of time.

> Also, did you imagine the page would update in real time?

yes and no :smile: real-time would be nice to have. But I would suggest first we just roll out a simple page that gives the merge queue status and refreshing the page for the latest state. That would also help us in evaluating if real-time is actually needed or not.

**Comment by chdsbd** [comment]
I released #617 to production, which fixes this issue. You can see active merge queues at the bottom of the main activity page on http://app.kodiakhq.com.

Let me know if you have questions or feedback.

**Comment by scriptnull** [comment]
Thanks :+1: That was fast :rocket: I will give it a try :slightly_smiling_face:
